prov/verbs/src/windows/verbs_nd_ov.c

```diff
@@ -313,29 +325,48 @@ void nd_cm_connect_ack(struct nd_event_base *base, DWORD bytes)
 		if (!cm_event)
 			cm_event = nd_allocate_cm_event(
 				id, RDMA_CM_EVENT_ESTABLISHED);
+		break;
 
+	case ND_CM_DISCONNECT:
+	case ND_CM_DISCONNECTED:
+		cm_event = nd_allocate_cm_event(id, RDMA_CM_EVENT_DISCONNECTED);
+		break;
+
+	default:
+		break;
+	}
+
+	VRB_DBG(FI_LOG_EQ, "EQ SET event:%p %p %d\n", cm_event->event.id,
+		cm_event->event.listen_id, cm_event->event.event);
+	dlistfd_insert_tail(&cm_event->entry, &ch_nd->q);
+
+	// The call to NotifyDisconnect needs to happen after we insert the
+	// CM event into the queue in order to avoid a race condition where
+	// a disconnect is virtually immediate and that callback interrupts
+	// this processing leading to an RDMA_CM_EVENT_DISCONNECTED coming
+	// before its associated RDMA_CM_EVENT_ESTABLISHED.
+	switch (event->state) {
```

Comment:
I don't understand what's happening here.  There's a switch statement above this that checks for event->state.  AFAICT, there's no change to event, then we have another switch statement here on event->state.  Why are there 2 switch statements?

Reply:
There are 2 switch statements because what happens in the first switch prepares cm_event->entry to be placed into ch_nd->q at line 341. The second switch makes a call to NotifyDisconnect and that call MUST happen after the event is inserted into the queue for the reason I tried to explain in the comment at lines 343-347.

Obviously I failed at making that clear. Do you have a suggestion for how I can make this more understandable?

Reply:
Please restructure to have a single switch statement.  We're talking about a single line of real code being added to the case statements.  Also, I'm guessing nd_allocate_cm_event() can fail, in which case we should handle that failure.

As a separate comment, it seems odd that we set the peer_event state to DISCONNECTED below as part of handling a connect ack.  That's non-intuitive.  It makes more sense to set the state after we received the notification of the disconnect.  As it is, it appears to be a predicted future state, not a current one as one would expect.